packages/flutter/lib/src/widgets/scroll_position_with_single_context.dart

```diff
@@ -0,0 +1,417 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/physics.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/scheduler.dart';
+import 'package:meta/meta.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'gesture_detector.dart';
+import 'scroll_activity.dart';
+import 'scroll_context.dart';
+import 'scroll_notification.dart';
+import 'scroll_physics.dart';
+import 'scroll_position.dart';
+
+/// A scroll position that manages its own scroll activities.
+///
+/// This class combines the [ScrollPosition] logic (which implements
+/// [ScrollPositionReadInterface]) with logic for using [ScrollActivity]
+/// objects. It provides a self-contained object for managing a single
+/// [Scrollable].
+///
+/// See also:
+///
+///  * [ScrollPosition], which defines the API for managing the scroll offset of
+///    a [Scrollable] viewport.
+///  * [ScrollView] and its subclasses such as [ListView], which use
+///    [ScrollIndependentPosition] to manage their scroll position.
+///  * [ScrollController], which can manipulate one or more [ScrollPosition]s,
+///    and which uses [ScrollIndependentPosition] as its default class for scroll
+///    positions.
+class ScrollPositionWithSingleContext extends ScrollPosition implements ScrollActivityDelegate {
+  /// Create a [ScrollPosition] object that manages its behavior using
+  /// [ScrollActivity] objects.
+  ///
+  /// The `initialPixels` argument can be null, but in that case it is
+  /// imperative that the value be set, using [correctPixels], as soon as
+  /// [applyNewDimensions] is invoked, before calling the inherited
+  /// implementation of that method.
+  ScrollPositionWithSingleContext({
+    @required ScrollPhysics physics,
+    @required this.context,
+    double initialPixels: 0.0,
+    ScrollPosition oldPosition,
+  }) : super(physics: physics, oldPosition: oldPosition) {
+    // If oldPosition is not null, the superclass will first call absorb(),
+    // which may set _pixels and _activity.
+    assert(physics != null);
+    assert(context != null);
+    assert(context.vsync != null);
+    if (pixels == null)
```

Comment:
maybe `&& initialPixels != null` for clarity (not that it really makes a difference in practice)